yarn-project/foundation/src/abi/function_selector.ts

```diff
@@ -0,0 +1,111 @@
+import { ABIParameter } from '@aztec/foundation/abi';
+import { toBigIntBE, toBufferBE } from '@aztec/foundation/bigint-buffer';
+import { keccak } from '@aztec/foundation/crypto';
+import { Fr } from '@aztec/foundation/fields';
+import { BufferReader } from '@aztec/foundation/serialize';
+
+/**
+ * A function selector is the first 4 bytes of the hash of a function signature.
+ */
+export class FunctionSelector {
+  /**
+   * The size of the hash in bytes.
+   */
+  public static SIZE = 4;
+
+  constructor(/** number representing the function selector */ public value: number) {
+    if (value > 2 ** (FunctionSelector.SIZE * 8) - 1) {
+      throw new Error(`Function selector must fit in ${FunctionSelector.SIZE} bytes.`);
+    }
+  }
+
+  /**
+   * Checks if the function selector is empty (all bytes are 0).
+   * @returns True if the function selector is empty (all bytes are 0).
+   */
+  public isEmpty(): boolean {
+    return this.value === 0;
+  }
+
+  /**
+   * Serialize as a buffer.
+   * @returns The buffer.
+   */
+  toBuffer(): Buffer {
+    return toBufferBE(BigInt(this.value), FunctionSelector.SIZE);
+  }
+
+  /**
+   * Serialize as a hex string.
+   * @returns The string.
+   */
+  toString(): string {
+    return this.toBuffer().toString('hex');
+  }
+
+  /**
+   * Checks if this function selector is equal to another.
+   * @param other - The other function selector.
+   * @returns True if the function selectors are equal.
+   */
+  equals(other: FunctionSelector): boolean {
+    return this.value === other.value;
+  }
+
+  /**
+   * Deserializes from a buffer or reader, corresponding to a write in cpp.
+   * @param buffer - Buffer  or BufferReader to read from.
+   * @returns The FunctionSelector.
+   */
+  static fromBuffer(buffer: Buffer | BufferReader): FunctionSelector {
+    const reader = BufferReader.asReader(buffer);
+    const value = Number(toBigIntBE(reader.readBytes(FunctionSelector.SIZE)));
+    return new FunctionSelector(value);
+  }
```

Comment:
I know we don't do it in any other `fromBuffer` method, but I'm thinking whether it'd make sense to validate if `buffer.length == SIZE` if the param is a buffer and not a reader.